### PR TITLE
[GH-20] triggerIfCalledFromOutside

### LINKED ISSUE #20: Convert temporarily ignored to file based triggerIfNotCalledFromWithinPackage
We pass the package name, so we can use that for testing for `vendor/{$packageName}` in the caller. If that matches, the deprecation should be suppressed.

This would avoid having to first ignore then trigger a deprecation, which might look weird.

This will replace and remove #18 (proposed in #16)

### PR DESCRIPTION
**Pull request by beberlei**
Fixes #20 and reverts #18

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,6 +80,18 @@ the message.
 );
 ```
 
+When you want to trigger a deprecation only when it is called by a function
+outside of the current package, but not trigger when the package itself is the cause,
```
Comment (greg0ire) [inline README.md:84]:
```suggestion
outside of the current package, but not trigger when the package itself is the caller,
```

File: README.md [hunk 2]
```diff
@@ -77,6 +80,18 @@ the message.
...
+then use:
+
+```php
+\Doctrine\Deprecations\Deprecation::triggerIfCalledFromOutside(
```
Comment (greg0ire) [inline README.md:88]:
```suggestion
\Doctrine\Deprecations\Deprecation::triggerOnExternalCall(
```

File: lib/Doctrine/Deprecations/Deprecation.php [hunk 3]
```diff
@@ -72,15 +70,49 @@ class Deprecation
...
+        } else {
+            self::$ignoredLinks[$link] = 1;
+        }
+
+        if (self::$deduplication === true && self::$ignoredLinks[$link] > 1) {
+            return;
+        }
+
+        // do not move this condition to the top, because we still want to
+        // count occcurences of deprecations even when we are not logging them.
```
Comment (greg0ire) [inline lib/Doctrine/Deprecations/Deprecation.php:84]:
```suggestion
        // count occurences of deprecations even when we are not logging them.
```

File: lib/Doctrine/Deprecations/Deprecation.php [hunk 4]
```diff
@@ -72,15 +70,49 @@ class Deprecation
...
+        $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
 
-            if (self::$temporarilyIgnoredLinks[$link] <= 0) {
-                unset(self::$temporarilyIgnoredLinks[$link]);
-            }
+        // "outside" means we assume that $package is currently installed as a
+        // dependency and the caller is not a file in that package.
+        // When $package is installed as a root package, then this deprecation
+        // is always ignored
+        if (strpos($backtrace[0]['file'], 'vendor/' . $package . '/') === false) {
```
Comment (greg0ire) [inline lib/Doctrine/Deprecations/Deprecation.php:111]: Apparently `vendor` can be configured: https://getcomposer.org/doc/06-config.md#vendor-dir

Here is a (hairy) way to detect it: https://github.com/symfony/phpunit-bridge/blob/f2985c4d17f5d2584c5bd66e13b2fd0e3a67416f/DeprecationErrorHandler/Deprecation.php#L312

But maybe that's overkill…

**Review by greg0ire — COMMENTED** [review]
Might be worth creating a directory with a fake app with fake vendors for test purposes. Example: https://github.com/symfony/symfony/tree/5.x/src/Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler

File: lib/Doctrine/Deprecations/Deprecation.php (hunk 4 above)
Comment (beberlei) [reply lib/Doctrine/Deprecations/Deprecation.php:111]: I know its not perfect, but its a cheap way to do the check vs the one in phpunit bridge that does not have to care for performance that much.

File: lib/Doctrine/Deprecations/Deprecation.php [hunk 5]
```diff
@@ -72,15 +70,49 @@ class Deprecation
...
+
+        self::delegateTriggerToBackend($message, $backtrace, $link, $package);
+    }
+
+    /**
+     * @param mixed $args
+     */
+    public static function triggerIfCalledFromOutside(string $package, string $link, string $message, ...$args): void
+    {
+        $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
```
Comment (morozov) [inline lib/Doctrine/Deprecations/Deprecation.php:105]: Unconditional usage of `debug_backtrace()` makes this method unnecessarily expensive even if logging is disabled.

File: lib/Doctrine/Deprecations/Deprecation.php [hunk 6]
```diff
@@ -72,15 +70,49 @@ class Deprecation
...
+            self::$ignoredLinks[$link]++;
+        } else {
+            self::$ignoredLinks[$link] = 1;
+        }
+
+        if (self::$deduplication === true && self::$ignoredLinks[$link] > 1) {
+            return;
+        }
+
+        // do not move this condition to the top, because we still want to
```
Comment (morozov) [inline lib/Doctrine/Deprecations/Deprecation.php:83]: I believe, in production, this method should return as early as possible.

Comment (beberlei) [reply lib/Doctrine/Deprecations/Deprecation.php:83]: Yes, i agree. 

I started that in the follow up https://github.com/doctrine/deprecations/pull/24 - though I have an idea to simplify that even more.

**Comment by nicolas-grekas** [comment]
This feature should not be required actually: a codebase should not call any of its own deprecated API ever. When internal calls to deprecated APIs cannot be avoided for legacy reasons, adding a quick internal API is going to be much faster and seamless (eg an extra arg to disable the deprecation.)

We achieved this in Symfony, so I'm pretty sure there is no need for this at all.

**Comment by morozov** [comment]
> a codebase should not call any of its own deprecated API ever.

This API may be an extension point. Not calling it is already a BC break.

**Comment by nicolas-grekas** [comment]
I agree, yet there is always a way to make this API internal. We do it when needed, it's work perfectly fine.

**Comment by morozov** [comment]
If you make it internal and stop calling, it's still a BC break because in a previous version it wasn't internal and was called. Could you point to an example of the approach you're talking about?

**Comment by nicolas-grekas** [comment]
Sure, check eg https://github.com/symfony/symfony/blob/abf16d0f1b42e7c289b6b1b57e427c0dbd35b861/src/Symfony/Component/HttpKernel/Kernel.php#L299-L303

**Comment by morozov** [comment]
> a codebase should not call any of its own deprecated API ever.

How does this align with this change?

https://github.com/symfony/symfony/commit/98ff750ee148a1c44032b67a669c9752930a28b9#diff-842527ea5582808ca82ff360aa1f7ceaa44584c9c4b4edd6c27381f62d834d3cR84

**Comment by nicolas-grekas** [comment]
The difference is that the code is self-aware, dealing internally with its own concerns. It doesn't leak this concern to a 3rd party lib that will needs to hack with the call stack to get past any API boundaries. In terms of perf also, it's negligible, compared to calling `debug_backtrace()`.

**Comment by morozov** [comment]
> The difference is that the code is self-aware, dealing internally with its own concerns. It doesn't leak this concern to a 3rd party lib [...]

Why is it a problem? All projects use 3rd-party libraries of other concerns.

> In terms of perf also, it's negligible, compared to calling `debug_backtrace()`.

In terms of perf, the approach in this library should be close to a no-op in production. In the rest of the environments, it's not a real concern.

Unlike `@trigger_error()` which only God knows how will behave depending on the runtime configuration.

> a codebase should not call any of its own deprecated API ever.

Does this argument still hold?

**Comment by nicolas-grekas** [comment]
I think it's a concern yes when there is no proper API involved. Hacking with the backtrace is doing black magic.

What do you mean by no-op? It's true I never benched the overhead of calling `debug_backtrace()`. Did you? Do you have any numbers?

About my argument, it holds, but I failed to communicate it appropriately. My fear is that this method will be abused to create broken deprecation layers. When deprecating something, authors should try as hard a possible to make their own codebase deprecation-free. This is an involving process. This method could be a convenient way to skip writing a proper BC layer.

**Comment by beberlei** [comment]
He is referring to the outstanding PR that checks for the mode==none, which will exit early in production before calling debug_backtrace

**Comment by beberlei** [comment]
@nicolas-grekas reimplementing the deprecation flags over and over is error prone and in case of some of our more funny PDO extended APIs with variable arguments not really possible. 

We tried the flags as you can see in my DBAL PR, but they make it much harder than this approach. We are encapsulating the problem for our use case and make it reusable. As a side benefit triggering deprecations stays deliberative and does not become imperative

**Comment by morozov** [comment]
> My fear is that this method will be abused to create broken deprecation layers [...] This method could be a convenient way to skip writing a proper BC layer.

IMO this fear is orthogonal to the implementation. In both cases, some calls to the deprecated methods are allowed. Although in this case, it's implemented by a very specific method which you can grep and see which exactly of your APIs are still used.

> Hacking with the backtrace is doing black magic.

So I assume, hacking with handling suppressed error messages and using non-existing method arguments is not?

**Comment by nicolas-grekas** [comment]
> Unlike @trigger_error() which only God knows how will behave depending on the runtime configuration.

Not "God", but also "not your business" - aka these are decoupled. You don't need to create defensive strategies against authors of userland error handlers. That's what we did in Symfony, and it just works well. Everybody does their part and things play nicely together.

> hacking with handling suppressed error messages and using non-existing method arguments is not?

This feels like a rhetorical question... In case it's not: `trigger_error()` + `@` are defined to be used like we do in Symfony. There is no hack here, this is just using the tools for what they've been designed for.

> I agree that the Symfony's approach is more fine-grained in the way that it allows callers to certify "I'm calling a deprecated method" rather than the deprecated method stating "I'm allowed to be called from within the library". But I'm not sure how essential it is at the moment.

Thanks for putting it that way, you've got it right. Not a minor thing to me, especially since you plan to release this for reuse. That's a significant foot gun.

**Pull request merged**